### ISSUE TITLE
Terraform wants to reach storage account static website endpoint even when none will be created

### ISSUE THREAD

**Issue by camilo-s** [hashicorp/terraform-provider-azurerm#20257]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.3.6

### AzureRM Provider Version

3.41.0

### Affected Resource(s)/Data Source(s)

azurerm_storage_account

### Terraform Configuration Files

```hcl
resource "azurerm_storage_account" "this" {
  name                            = local.storage_account_name
  resource_group_name             = var.resource_group_name
  location                        = var.location
  account_tier                    = var.account_tier
  account_replication_type        = var.account_replication_type
  access_tier                     = var.access_tier
  min_tls_version                 = var.min_tls_version
  allow_nested_items_to_be_public = false
  is_hns_enabled                  = var.is_hns_enabled

  network_rules {
    default_action = "Deny"
    bypass         = ["AzureServices"]
  }
}

resource "azurerm_private_endpoint" "this" {
  name                = "pe-${azurerm_storage_account.this.name}"
  resource_group_name = var.resource_group_name
  location            = var.location
  subnet_id           = var.subnet_id

  private_service_connection {
    name                           = "psc-${azurerm_storage_account.this.name}"
    is_manual_connection           = false
    private_connection_resource_id = azurerm_storage_account.this.id
    subresource_names              = ["blob"]
  }
}

data "azurerm_private_dns_zone" "this" {

  name                = "privatelink.blob.core.windows.net"
  resource_group_name = var.private_dns_resource_group_name

  provider = azurerm.private_dns_zone # Provider targeting subscription with private DNS Zone
}

resource "azurerm_private_dns_a_record" "this" {

  name                = azurerm_storage_account.this.name
  zone_name           = data.azurerm_private_dns_zone.this.name
  resource_group_name = data.azurerm_private_dns_zone.this.resource_group_name
  ttl                 = 3600

  records = [
    azurerm_private_endpoint.this[each.key].private_service_connection[0].private_ip_address
  ]

  provider = azurerm.private_dns_zone # Provider targeting subscription with private DNS Zone

}
```

### Debug Output/Panic Output

https://gist.github.com/camilo-s/90e4fa4dbe90ae3cefee27f69fe4bba5

### Expected Behaviour

The storage account and its private endpoint get deployed normally.

### Actual Behaviour

The following error occurs while creating the Storage Account
```
Error: reading static website for AzureRM Storage Account "stdataplatformdevb8b849": accounts.Client#GetServiceProperties: Failure sending request: StatusCode=0 -- Original Error: context deadline exceeded
with module.data_platform.module.data_lake.module.storage_account.azurerm_storage_account.this
on .terraform/modules/data_platform.data_lake.storage_account/main.tf line 115, in resource "azurerm_storage_account" "this":
resource "azurerm_storage_account" "this" {
```

### Steps to Reproduce

1. `terraform init`
2. `terraform plan`
3. `terraform apply`

### Important Factoids

The DNS A Record for the private endpoint is created separately via Terraform. Terraform runs on a Terraform Enterprise Agent inside a VNet; it does have access to the target subnet that will host the storage account. This errors occurs randomly (e.g. TF trying to reach unintended queue endpoint), the same code has also succeeded under the same circumstances. The TFE agent uses an https Proxy in general, but is configured to bypass it when communicating the intended Azure Blob endpoint.

**Comment by Fodsuk** [comment]
this commit looks to be related - https://github.com/hashicorp/terraform-provider-azurerm/pull/19062/files

**Comment by camilo-s** [comment]
Here's a curious factoid: I've downgraded the `azurerm` provider version to `3.37.0` (last time my deployment was known to work) and it does work smoothly having the proxy configured to use the storage account private endpoints.

**Comment by Fodsuk** [comment]
@magodo - could this be related to your change here https://github.com/hashicorp/terraform-provider-azurerm/pull/19062/files

**Comment by magodo** [comment]
From the error message, the request towards the blob endpoint: `https://<accountname>.blob.core.windows.net` hangs somehow. This should be not introduced by #19062.

The fact is that terraform will always try to read the static website properties (via the blob data plane ndpoint) during read, as long as the `account_kind` supports that feature:

https://github.com/hashicorp/terraform-provider-azurerm/blob/4ce0783b21fd4c07c4160aae37e910a5a8708870/internal/services/storage/storage_account_resource.go#L2210-L2230

This issue is more likely related to #13070, where we observed sometimes the data plane endpoint are not ready after the storage account is created.

**Comment by camilo-s** [comment]
In my case, Terraform even reports the storage account creation as failed (as opposed to the returned 200 mentioned in https://github.com/hashicorp/terraform-provider-azurerm/issues/13070), though it get's deployed (as seen in the Azure Portal). 

Any further attempts at interacting with the storage account via Terraform, like at the `terraform plan` stage, fail while bypassing the proxy (understandably so, because no Private Endpoint could be created and the storage account has no network exceptions).

So the only workaround with `azurerm` 3.41.0 is to juggle the storage account in an out of the `no_proxy` env, whereas `azurerm` 3.37.0 is able to handle this end to end.

**Comment by magodo** [comment]
@camilo-s Could you please have a minimal reproduce of this issue and provide the debug log. I've tried to look into the gist you provided, but I'm not sure where does the issue occur (only three matches when searching for `storage_account`).

**Comment by camilo-s** [comment]
Hello @magodo, I've updated the gist to include the full debug logs (with masking).

**Comment by magodo** [comment]
@camilo-s The log still contains no useful info.. Can you ensure you exported env var `TF_LOG_PROVIDER=debug`?

**Comment by magodo** [comment]
@camilo-s Sorry I didn't realized the file is truncated. With exploring the full file, I noticed that the last call in the log for the affected storage account is a `GET` towards: `https://management.azure.com/subscriptions/#####/resourceGroups/rg-##masked##-germanywestcentral-004/         providers/Microsoft.Storage/storageAccounts/storageaccountmasked#####1/fileServices/default?api-version=2021-09-01`, which is successful. This corresponds to:

https://github.com/hashicorp/terraform-provider-azurerm/blob/f7ba194abc06bf5532d4e12fe13cf5c774d6cf76/internal/services/storage/storage_account_resource.go#L2197-L2208

As your storage's kind is StorageV2, the provider will then move on to reading for static website information from the blob data plane endpoint (e.g. https://unlikely23exst2acctmgd6.blob.core.windows.net/?comp=properties&restype=service):

https://github.com/hashicorp/terraform-provider-azurerm/blob/f7ba194abc06bf5532d4e12fe13cf5c774d6cf76/internal/services/storage/storage_account_resource.go#L2210-L2230

Unfortunately, currently this API call won't be logged as the underlying SDK (another data plane SDK that is different than the management plane ones) seems not support that yet. That is why you won't see this API call from log. But if you have a proxy, you'll see the API anyway.

It is worth mentioning that, prior to calling the `GET` to the file service, the provider calls a `GET` to the queue endpoint if current storage account is a standard StorageV2 one (just as yours), which is a data plane URI (e.g. https://unlikely23exst2acctmgd6.queue.core.windows.net/?comp=properties&restype=service):

https://github.com/hashicorp/terraform-provider-azurerm/blob/f7ba194abc06bf5532d4e12fe13cf5c774d6cf76/internal/services/storage/storage_account_resource.go#L2181-L2195

Looks like above call is successful in your case, which indicates that the network should be fine at least for the queue endpoint. So the problem is why the blob data plane endpoint is not reachable. My gut feeling is that something went wrong at service side when setting it up during creation (especially if you are recreating a new one but with the same name).

**Comment by camilo-s** [comment]
Hi @magodo, I'm using `random_suffix` resources for resource naming, so I don't believe it is about attempting to recreate resources with the same name. 

On a separate that doesn't explain why my deployment works with `azurerm` provider 3.37.0 but not with 3.41.0 (with all other configurations and setup identical)

**Comment by magodo** [comment]
@camilo-s By looking into the history of this resource between v3.37.0 and v3.41.0, there is no suspicious commit found. I think keep the problematic environment and submit a support ticket would be much easier to see which part went wrong? If the issue is at the provider side, then we can continue looking into how to fix that.

**Comment by p4ranoidandro1d** [comment]
On my side, using 3.40.0

Storage Account now fails because provider tries to talk to queue endpoint though i dont use this.
Our context is :

 - We create a storage account : 
   - with AccessKey disabled by default (and by policy)
   - Default to oauth authent is true

Our terraform code only applies network rules once the storage account is created (with  "azurerm_storage_account_network_rules" resource)

Right now, error message is 
```
│ Error: reading queue properties for AzureRM Storage Account "stcorpgrafanamimir": queues.Client#GetServiceProperties: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="KeyBasedAuthenticationNotPermitted" Message="Key based authentication is not permitted on this storage account.\nRequestId:c3a9a108-3033-0018-2ee6-3add8d000000\nTime:2023-02-07T11:21:47.8988886Z"
```

Digging up how i can workaround this

**Comment by camilo-s** [comment]
@p4ranoidandro1d have you tried setting the [`storage_use_azuread`](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs#storage_use_azuread) flag to `true`?
https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs#storage_use_azuread

**Comment by p4ranoidandro1d** [comment]
> 

Well, that fixes it. Thanks !

**Comment by amira1o-edf** [comment]
experiencing similar issue. In my case, the error occurs when performing a terraform plan.

**Comment by magodo** [comment]
@amira1o-edf Could you please provide a minimal reproduce so that I can get the error at my side? Also, do you mean a lower version of provider won't get this error?

**Comment by amira1o-edf** [comment]
### AzureRM Provider Version
3.40.0

### Affected Resource(s)/Data Source(s)
azurerm_storage_account

### Terraform Configuration Files
```terraform
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "~>3.40.0"
    }
  }
  backend "azurerm" {}
}

resource "azurerm_storage_account" "example_sa" {
  name                      = local.storage_account_name
  account_tier              = "Standard"
  account_kind              = "StorageV2"
  resource_group_name       = data.azurerm_resource_group.rg.name
  account_replication_type  = "GRS"
  location                  = data.azurerm_resource_group.rg.location
  min_tls_version           = "TLS1_2"
  enable_https_traffic_only = true

  queue_properties {
    logging {
      delete                = true
      read                  = true
      write                 = true
      version               = "1.0"
      retention_policy_days = 10
    }
    hour_metrics {
      enabled               = true
      include_apis          = true
      version               = "1.0"
      retention_policy_days = 10
    }
    minute_metrics {
      enabled               = true
      include_apis          = true
      version               = "1.0"
      retention_policy_days = 10
    }
  }

  blob_properties {
    versioning_enabled = true

    delete_retention_policy {
      days = 7
    }
  }

  tags = merge(module.tags.tags, { "name" = local.storage_account_name })
}

resource "azurerm_storage_encryption_scope" "example_sa_encryption" {
  depends_on         = [azurerm_storage_account.example_sa]
  name               = "microsoftmanaged"
  storage_account_id = azurerm_storage_account.example_sa.id
  source             = "Microsoft.Storage"
}
```

### Expected Behaviour
The existing storage account get refreshed normally.

### Actual Behaviour
The following error occurs after 5 minutes while refreshing the existing storage account 
```terraform
azurerm_storage_account.example_sa: Refreshing state... [id=/subscriptions/######/resourceGroups/example_rg/providers/Microsoft.Storage/storageAccounts/example_sa_name]


╷
│ Error: reading static website for AzureRM Storage Account "example_sa_name": accounts.Client#GetServiceProperties: Failure sending request: StatusCode=0 -- Original Error: context deadline exceeded
│ 
│   with azurerm_storage_account.example_sa,
│   on storage_account.tf line 1, in resource "azurerm_storage_account" "example_sa":
│    1: resource "azurerm_storage_account" "example_sa" {
│ 
╵
##[error]Error: Terraform Plan failed with exit code: 1

```

### Steps to Reproduce
1. ```terraform init```
2. ```terraform plan```

**Comment by magodo** [comment]
@amira1o-edf I can `terraform apply` then `terraform plan` above config with no error via the latest provider v3.43.0, locally.

**Comment by alex-slynko** [comment]
Hi
I can reproduce this issue with Terraform running from Azure in a separate vNet and subscription.

To do this I create storage account and Azure Data Explorer.
Then I add the Kusto private endpoint to storage account, approve it and see the error.

```
resource "azurerm_storage_account" "data" {
    name                      = local.account_name
      resource_group_name       = azurerm_resource_group.rg.name
      location                  = azurerm_resource_group.rg.location
      account_kind              = "StorageV2"
      account_tier              = "Standard"
      account_replication_type  = "GRS"
      is_hns_enabled            = true
      enable_https_traffic_only = true
    }

  resource "azurerm_kusto_cluster_managed_private_endpoint" "blob_endpoint" {
    name                         = "${data.azurerm_kusto_cluster.kusto.name}-data-test2-blob"
    resource_group_name          = data.azurerm_kusto_cluster.kusto.resource_group_name
    cluster_name                 = data.azurerm_kusto_cluster.kusto.name
    private_link_resource_id     = azurerm_storage_account.data.id
    private_link_resource_region = "eastus"
    group_id                     = "blob"
    request_message              = "Auto Generated by Terraform, Please Approve"
  }
```

The Azure Data Factory was created with the default values.

**Comment by slzmruepp** [comment]
Same issue here with  version = ">= 3.45.0"

**Comment by joshua-hancox** [comment]
I also have the same issue. 
Terraform Version:  v1.3.5

This is for an existing storage account created with Terraform.

When upgrading to provider version v3.40.0 (or later) I see the error:

> │ Error: reading static website for AzureRM Storage Account "xxxxxxxxxxx": accounts.Client#GetServiceProperties: Failure sending request: StatusCode=0 -- Original Error: context deadline exceeded
> │
> │   with azurerm_storage_account.storage_account.blah,
> │   on .terraform/modules/ukw_storage_account_1/main.tf line 1, in resource "azurerm_storage_account" "storage_account":
> │    1: resource "azurerm_storage_account" "storage_account" {

If I use provider version v3.39.1 (or before) I can run my `terraform plan` successfully.

Surely this is too much of a coincidence not to be related to https://github.com/hashicorp/terraform-provider-azurerm/pull/19062/files (contained within the same release, makes a change related to the error received)

I am also using private endpoints with my storage account, and we also create an A record for the blob endpoint separately via terraform. I assume that the issue here comes because there is a different endpoint for sites? And we do not create that A record. I don't think the provider should force you to create extra resources though?

cc: @magodo @camilo-s

**Comment by magodo** [comment]
@joshua-hancox Thank you for providing more context here. #19062 hide the context timeout error happens when accessing the blob endpoint (instead of the static web site). You are right that *the provider should NOT force you to create extra resources* in a minor version update. I'll revert the change for the static web site for now. Just out of curiosity, what is the config of PE for your setup? Have you set the "blob" in its `subresource_names`?

**Comment by timja** [comment]
This is our private endpoint which is having the same issue:

```terraform
resource "azurerm_private_endpoint" "sa" {
  name                = "${var.product}sa-pe-${var.env}"
  resource_group_name = azurerm_resource_group.rg.name
  location            = azurerm_resource_group.rg.location
  subnet_id           = azurerm_subnet.endpoint_subnet.id

  private_service_connection {
    name                           = "${var.product}sa-psc-${var.env}"
    is_manual_connection           = false
    private_connection_resource_id = module.sa_storage_account.storageaccount_id
    subresource_names              = ["blob"]
  }
  tags = var.common_tags
}
```

**Comment by joshua-hancox** [comment]
Thankyou @magodo - yes we do have "blob" included in the sub_resource names, similar to timja's setup above.

```hcl
resource "azurerm_private_endpoint" "example" {

  for_each = var.subresources

  name                = "example"
  location            = "uksouth"
  resource_group_name = "example_rg_name"
  subnet_id           = "example_subnet_id"

  private_service_connection {
    name                           = "example"
    is_manual_connection           = false
    private_connection_resource_id = azurerm_storage_account.example.id
    subresource_names              = ["blob"]
  }

  private_dns_zone_group {
    name                 = "example"
    private_dns_zone_ids = [private_dns_zone.example.id]
  }
}
```

**Comment by tombuildsstuff** [comment]
@magodo looking through here this appears to be another instance of #2977, where the only means of obtaining this information is through the Data Plane API - hiding the error in #20727 isn't going to do much to help here tbh?

Presuming that the Data Plane API is available via Private Link, it'd be better to query that using that endpoint rather than the public one, if we can detect that Private Link is being used?

In either case as this is another instance of #2977, it's probably worth closing this issue in favour of that one, so that we can keep this consolidated in one place?

**Comment by magodo** [comment]
@tombuildsstuff You're right that #20727 won't fix the underlying issue, while it at least unblock users at some extent, especially when they never cares about the static web site related configs.

Regarding to detect whether the provisioner is in a subnet that has private endpoint enabled, per https://learn.microsoft.com/en-us/azure/storage/common/storage-private-endpoints#dns-changes-for-private-endpoints, it doesn't seem to make a difference as the DNS resolution starts from the same data plane endpoint, which will then resolve to the privatelink sub domain name, and then depending on whether you are in the same vnet, it resolve to different name servers.

I've also made a test via the key vault that I have the PE enabled and created the private dns zone, but I just didn't add the record for the PE in the dns zone. Then
- If I access it from outside of the vnet that the PE is configured, it still get's resolved (via the public name server, same as storage account's doc said)
- If I access it from within the vnet, it failed to resolve the ip (only a CNAME record pointing to the privatelink sub domain), which is as expected

So just want to ensure @joshua-hancox and @timja run into the error with private dns zone correctly setup the record for PE's private ip? Also, did you run into this issue when recreating a same named strorage account (as in this case, the DNS TTL might still point to the old one, and that one has been destroyed).

**Comment by camilo-s** [comment]
@magodo I can confirm that the Private DNS Zone configuration in my case was correct. 

As stated in the OP, I was able to interact normally against the blob private endpoint prior to provider version 3.39.0 (that is, 3.37.0 which I was using at the time), both from Terraform and as an external consumer of the endpoint.

**Comment by jan-hudec** [comment]
I'm testing problems in somewhat related configuration (IoT Hub access through private links) and … for the test I added my client IP into the network rules, which allowed me to create a blob container, but I am still getting this error on refresh.

That suggests that it is a different endpoint than the `blob.core.windows.net` one that is the problem here, and that it probably is a problem that this endpoint gets blocked from some other Azure service, not from direct access.

**Comment by magodo** [comment]
> @magodo I can confirm that the Private DNS Zone configuration in my case was correct.
> 
> As stated in the OP, I was able to interact normally against the blob private endpoint prior to provider version 3.39.0 (that is, 3.37.0 which I was using at the time), both from Terraform and as an external consumer of the endpoint.

That said, the reason why previous version *works* doesn't really mean it works, which is because the error is swallowed by the provider.

**Comment by pierluca** [comment]
I've also gotten bitten by this on a project, it's completely blocked one of the organization's Terraform Cloud workspaces. 
I went through this thread, but I can't exactly figure out if there's a practical workaround to unblock the situation ?

Our setup is that we only use blob storage and have configured a private endpoint to access it, with a SAS key.
There's no static website involved.

**Comment by jan-hudec** [comment]
I've done some testing on this.

1. The request stops working when a private endpoint is created for the storage account's `blob` subresource even though public network access was not restricted in the process.
2. The request that does not work is indeed to *saname*`.blob.core.windows.net`. It is also failing in the portal from the “static website” tab.
3. The behaviour is the same when I add the private endpoint—without restricting public access—from the portal, so there seems to be a bug on the Azure side. Terraform not being able to ignore that failure makes it much bigger problem for terraform users though—in the portal you don't need that tab for anything else.
4. I tried also creating a private endpoint for the `web` subresource, but it did not help.

On a side-note, in the browser the connection immediately breaks, but the provider waits for five minutes; it looks like the connection breaking may not be reported properly in the go client.

**Comment by magodo** [comment]
@jan-hudec Did you enabled private dns zone? If that is the case, without setting the A recrod for the PE, you are not able to access the blob endpoint. Otherwise, the public network access should be able to access the blob endpoint (try the dig web interface to resolve the dns).

**Comment by jan-hudec** [comment]
@magodo, yes, I did, but nothing was running from inside the network, so it should have still been using the public endpoint … which does not seem to exist at all.

**Comment by thllxb** [comment]
> @magodo looking through here this appears to be another instance of #2977, where the only means of obtaining this information is through the Data Plane API - hiding the error in #20727 isn't going to do much to help here tbh?
> 
> Presuming that the Data Plane API is available via Private Link, it'd be better to query that using that endpoint rather than the public one, if we can detect that Private Link is being used?
> 
> In either case as this is another instance of #2977, it's probably worth closing this issue in favour of that one, so that we can keep this consolidated in one place?

I think the [PR](https://github.com/hashicorp/terraform-provider-azurerm/pull/20727) from  @magodo is quite important to many of us. It does unblock many of us.

**Comment by magodo** [comment]
@tombuildsstuff Sure, I'll try to reach out them. Meanwhile, I've append this new gap to the gap list at: https://github.com/hashicorp/terraform-provider-azurerm/pull/14220#issuecomment-998397558.

**Comment by magodo** [comment]
@camilo-s I don't think private link is totally blocked, as long as you have the correct setup.

**Comment by magodo** [comment]
@joshua-hancox When you run `nslookup` and `telnet`, are you in the same machine that is running terraform?

**Comment by joshua-hancox** [comment]
Yes it's from the same machine.

**Comment by jan-hudec** [comment]
> Would you guys mind indicating in the `azurerm_storage_account` documentation that Private Link isn't currently supported with a storage account as @tombuildsstuff mentioned?

Private links are supported as long as you run terraform from a place that can properly see them (i.e. either on a VM in the vnet where the private endpoint is, or with a VPN to that vnet), and have the Azure DNS set up. The should also work as long as you are not connected to any Azure vnet. But they are broken if you are connected to a *different* vnet, because then the Azure DNS will try to direct you to the private endpoints you don't see even if you would be allowed to use the public endpoint.

**Comment by camilo-s** [comment]
@magodo and @jan-hudec my setup is a bit involved, but I know it worked with private endpoints because it had properly done so up to AzureRM provider 3.39 (slightly similar to that of joshua-hancox' above I guess):
- Terraform runs on an Azure DevOps self-hosted agent installed in a VM deployed in a spoke vnet peered with a hub vnet.
- An internet Proxy is bypassed for Azure resources that are meant to be reached via private endpoint (e.g. Storage Accounts, Key Vaults, etc.), communication with the resource manager API and other internet endpoints proceeds through the proxy.
- The spoke vnet uses custom DNS servers which forward lookups to Private DNS zones linked to the hub vnet. 
- I create the blob private endpoint and then its DNS A record manually in the above mentioned blob private DNZ zone.

The Terraform agent therefore _sees_ the private endpoints.

My understanding from @tombuildsstuff's comment above and on #1906, #2977 and https://github.com/hashicorp/terraform-provider-azurerm/pull/20727#issuecomment-1461359812  is that starting from AzureRM 3.40, Terraform no longer silently consumes some 404 error when reading static website properties from a storage account, which it had before. So it cannot faithfully map a storage account's configuration to the state file via a blob private endpoint, which effectively makes deploying storage accounts with AzureRM>3.40 and only blob private endpoints unfeasible.

I know it's a rather contrived scenario to be limited to interact with Azure only via the resource manager and private endpoints, but I'd still appreciate an indication of this limitation in the documentation.

**Comment by philippfrenzel** [comment]
Hi,

in our project we are short in available ip's for private endpoints - so we can't just create a blob private endpoint to the storage to avoid the error message.

Is their a setting/configuration that allows us to tell terraform to not check for the static website?!

Thanks Philipp

**Comment by magodo** [comment]
@philippfrenzel Not something I'm aware of, except you can workaround this by using azapi provider.

**Comment by philippfrenzel** [comment]
@magodo thanks for the quick answer - and posting out azapi provider - so we need to check all options and hopefully find an solution

**Comment by riemers** [comment]
Forgive me if this was already mentioned (or not) but i had the same issue that it complains about a 404. This is before it gets to the part to create the PLE. From my understanding above if you have a PLE (eventually) it will still work. The weird part is that it didn't work without PLE and i gave up. Destroyed everything and then tried again the next day and all worked fine (SA & PLE) So the question is, what causes this? Some kind of dns/cache that needs to expire? How do i know i can try again?

**Comment by viral1701** [comment]
We are having the same issue using a private agent. I tried to switch to `azapi provider` but was still hit with the same issue. The region I am trying to deploy the storage account with a private endpoint is North Europe.

**Comment by aquibchiniwala** [comment]
Same issue..terraform works fine from local machine but when running on Self Hosted VM Agent it start to route privately somewhere and fails. If I connect the Vnet where the private endpoint of the Storage account is hosted to the Hub Vnet where all my DNS resolutions are done..it works.

> @magodo and @jan-hudec my setup is a bit involved, but I know it worked with private endpoints because it had properly done so up to AzureRM provider 3.39 (slightly similar to that of joshua-hancox' above I guess):
> 
> * Terraform runs on an Azure DevOps self-hosted agent installed in a VM deployed in a spoke vnet peered with a hub vnet.
> * An internet Proxy is bypassed for Azure resources that are meant to be reached via private endpoint (e.g. Storage Accounts, Key Vaults, etc.), communication with the resource manager API and other internet endpoints proceeds through the proxy.
> * The spoke vnet uses custom DNS servers which forward lookups to Private DNS zones linked to the hub vnet.
> * I create the blob private endpoint and then its DNS A record manually in the above mentioned blob private DNZ zone.
> 
> The Terraform agent therefore _sees_ the private endpoints.
> 
> My understanding from @tombuildsstuff's comment above and on #19062, #2977 and [#20727 (comment)](https://github.com/hashicorp/terraform-provider-azurerm/pull/20727#issuecomment-1461359812) is that starting from AzureRM 3.40, Terraform no longer silently consumes some 404 error when reading static website properties from a storage account, which it had before. So it cannot faithfully map a storage account's configuration to the state file via a blob private endpoint, which effectively makes deploying storage accounts with AzureRM>3.40 and only blob private endpoints unfeasible.
> 
> I know it's a rather contrived scenario to be limited to interact with Azure only via the resource manager and private endpoints, but I'd still appreciate an indication of this limitation in the documentation.

Same issue..terraform works fine from local machine but when running on Self Hosted VM Agent it start to route privately somewhere and fails. If I connect the Vnet where the private endpoint of the Storage account is hosted to the Hub Vnet where all my DNS resolutions are done..it works.

**Comment by tombuildsstuff** [comment]
👋  hey folks

Since there's been a number of usage questions since this issue's been closed, I've marked those as `off-topic` to [ensure that this comment is visible, since this behaviour is currently working as intended](https://github.com/hashicorp/terraform-provider-azurerm/issues/20257#issuecomment-1461352568) - and the long term fix for this will be addressed by #2977.

As the original issue has been clarified and this has been closed for a couple of months now, I'm going to lock this issue for the moment - should you have further usage questions I'd [direct you towards the Community Discuss forum](https://discuss.hashicorp.com).

Thanks!
